### PR TITLE
Message and stream level data  and accept  mime

### PR DESCRIPTION
**Pull request by rudy2steiner**
At present there can be only one data MIME for the every request/stream on the connection, 
Supporting content-type and accept header like http in this pr.
Implementation of  issue #26379

Class MimeTypeMetadataCodec will be delete once  rsocket-java v1.1.1 release

### CODE REVIEW

**Comment by rudy2steiner** [comment]
@rstoyanchev PTAL

**Review by rstoyanchev — CHANGES_REQUESTED** [review]
Some high level feedback. 

To start, the `MetadataMimeCodec` should be in RSocket-Java, not here. I've created https://github.com/rsocket/rsocket-java/issues/996 for that.

On the side of RSocketRequester, I can see potentially methods to set the mimeTypes or acceptMimeTypes for a request but no need more generally to introduce the concept of "headers".

Some classes, like `AbstractEncoderMethodReturnValueHandler.java`, are intended to be more generally used and this is why they are not in the `rsocket` package. Such classes cannot have RSocket dependencies.

**Comment by rudy2steiner** [comment]
> Some classes, like `AbstractEncoderMethodReturnValueHandler.java`, are intended to be more generally used and this is why they are not in the `rsocket` package. Such classes cannot have RSocket dependencies.

I agree with MetadataMimeCodec should be in Rsocket-java, and do you have more specific feedbacks how to implement contentType and acceptMimes .

**Comment by rudy2steiner** [comment]
@rstoyanchev thanks for your feedbacks, updated!

File: spring-messaging/src/main/java/org/springframework/messaging/rsocket/annotation/support/MessagingRSocket.java [hunk 1]
```diff
@@ -203,16 +204,32 @@ private MessageHeaders createHeaders(
...
+	 **/
+	@SuppressWarnings({ "unchecked" })
+	public MimeType getMessageContentType(MessageHeaderAccessor headers){
+		Object messageContentType  = headers.getHeader(MessageHeaders.CONTENT_TYPE);
+		if (messageContentType instanceof List) {
+			List<Object> types = (List<Object>)messageContentType;
+			return types.size() == 0 ? this.dataMimeType : (MimeType) types.get(0);
+		}
+		return this.dataMimeType;
+	}
```
Comment (rstoyanchev) [inline spring-messaging/src/main/java/org/springframework/messaging/rsocket/annotation/support/MessagingRSocket.java:233]: Likewise here, when accessed `CONTENT_TYPE` should be expected to be a single `MimeType`.

File: spring-messaging/src/main/java/org/springframework/messaging/rsocket/annotation/support/RSocketPayloadReturnValueHandler.java [hunk 2]
```diff
@@ -50,22 +56,56 @@ public class RSocketPayloadReturnValueHandler extends AbstractEncoderMethodRetur
...
 
 	@Override
 	protected Mono<Void> handleEncodedContent(
 			Flux<DataBuffer> encodedContent, MethodParameter returnType, Message<?> message) {
 
 		AtomicReference<Flux<Payload>> responseRef = getResponseReference(message);
+		AtomicReference<MimeType> contentTypeAtomic = getResponseContentType(message);
 		Assert.notNull(responseRef, "Missing '" + RESPONSE_HEADER + "'");
-		responseRef.set(encodedContent.map(PayloadUtils::createPayload));
+		responseRef.set(encodedContent.switchOnFirst((signal, inner) -> {
```
Comment (rstoyanchev) [inline spring-messaging/src/main/java/org/springframework/messaging/rsocket/annotation/support/RSocketPayloadReturnValueHandler.java:78]: We shouldn't automatically use `switchOnFirst`. If no content negotiation took place, and the application is using the connection-wide `dataMimeType`, there is no need to use this operator.

File: spring-messaging/src/main/java/org/springframework/messaging/rsocket/DefaultRSocketRequester.java [hunk 3]
```diff
@@ -310,12 +349,27 @@ private <T> Flux<T> retrieveFlux(ResolvableType elementType) {
...
 					(T) decoder.decode(dataBuffer, elementType, dataMimeType, EMPTY_HINTS));
 		}
 
+		@SuppressWarnings("unchecked")
+		public <T> Flux<T> decodeFlux(Payload firstPayload, Flux<Payload> payloads,ResolvableType elementType){
+			MimeType contentMime =  messageContentType(firstPayload);
+			Decoder<?> decoder = strategies.decoder(elementType, contentMime);
+			return payloads.map(this::retainDataAndReleasePayload).map(dataBuffer ->
+					(T) decoder.decode(dataBuffer, elementType, contentMime, EMPTY_HINTS));
+		}
```
Comment (rstoyanchev) [inline spring-messaging/src/main/java/org/springframework/messaging/rsocket/DefaultRSocketRequester.java:371]: Likewise on the requester side, we should be aware if a request specific content type is being used, and if not then avoid using the `switchOnFirst` operator.

File: spring-messaging/src/main/java/org/springframework/messaging/handler/invocation/reactive/HandlerMethodReturnValueHandler.java [hunk 4]
```diff
@@ -33,7 +33,10 @@ public interface HandlerMethodReturnValueHandler {
 
 	/** Header containing a DataBufferFactory for use in return value handling. */
 	String DATA_BUFFER_FACTORY_HEADER = "dataBufferFactory";
-
+	/**
+	 * Message header name of response content type.
+	 **/
+	String RESPONSE_CONTENT_TYPE = "responseContentType";
```
Comment (rstoyanchev) [inline spring-messaging/src/main/java/org/springframework/messaging/handler/invocation/reactive/HandlerMethodReturnValueHandler.java:39]: It's not clear what the purpose of this header is. It seems to be set as an empty AtomicReference from `MessagingRSocket` in the beginning, and then its value is set from `AbstractEncoderMethodReturnValueHandler`, but I don't see anything making use of it.

**Review by rstoyanchev — CHANGES_REQUESTED** [review]
@rudy2steiner, I've reviewed the PR and added a few comments.

**Comment by rstoyanchev** [comment]
Closing for now as the PR needs significant updates, but we have #26379 to keep track of the request.

**Pull request closed without merging**